Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  WebSocketProviderMock.swift
+//  StreamChatClientTests
+//
+//  Created by Alexey Bukhtin on 01/05/2020.
+//  Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import XCTest
+@testable import StreamChatClient
+
+final class WebSocketProviderMock: WebSocketProvider {
+    let request: URLRequest
+    var isConnected: Bool = false
+    let callbackQueue: DispatchQueue
+    weak var delegate: WebSocketProviderDelegate?
+    private var connectionId: String?
+    
+    var failNextConnectCount = 0
+    let timeout = 3
+    private var timer: DispatchSourceTimer?
+    
+    init(request: URLRequest, callbackQueue: DispatchQueue) {
+        self.request = request
+        self.callbackQueue = callbackQueue
+    }
+    
+    func connect() {
+        if failNextConnectCount > 0 {
+            failNextConnectCount -= 1
+            
+            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
```

Comment:
I get the idea that you want to simulate async delegate callback but do we need to wait? Wouldn't simple `DispatchQueue.main.async` be enough?